### ISSUE TITLE
Enhancement - Antlr5 - roadmap / explorations / machine translation using neural nets

### ISSUE THREAD

**Issue by johndpope** [antlr/antlr4#1866]
I've been reviewing tensorflow recently and wanted to share this idea. It is most likely not suitable for this repo/ but not sure a better place to log this.

At the essence, think of one day translating code from one language to another in as simple a way as using google translate. This is actually very doable today using tensorflow + syntaxnet.

What follows could be one approach. Although more recent developments in dragnn could probably be factored in. Dragnn at its core builds its own gramma files. (Haven't fully got my head around how this could be seeded)

Consider that, 
Given a gramma file 
Some how / programmatically generate valid code in multiple languages. (Of use precaanned ones from Wikipedia )
Parse the AST in a super language (swift)
Train the net on this AST representation(s)
Given this AST in this language the corresponding AST in this other language is....
Use DCGAN to forge valid /compilable code.

Consider that this trained model would be thrown away each month with competing models.

**Comment by KvanTTT** [comment]
Why do you miss in ANTLR 4 for implementing such translator?

**Comment by johndpope** [comment]
In all honesty, it's an unknown. Tackling this problem alone is not my intention. But consider having support of industry to help. Perhaps a competition to translate the code would be appropriate. This way it would flesh out problem to find out what's missing. It could be sponsored by IBM / google / Microsoft / nvidia / intel. 

The training data is important / but being able to programmatically generate code maybe critical to feed back into training.

**Comment by linonetwo** [comment]
Is there a tool that can write CFG for some language X, then generate a code generator, that can translate AST that generated by Antlr4, to language X? Currently?

**Comment by KvanTTT** [comment]
We (@PositiveTechnologies) use a unified AST (UST) in our open source Pattern Matching Engine [PT.PM](https://github.com/PositiveTechnologies/PT.PM). UST obtained by converting an ANTLR parse tree which obtained from the parser and thus from the grammar.

Also we are developing a new proprietary engine for analyzing data flows on unified AST. For this UST being converted to CFG, to PDG, and to combined representation CPG (UST + CFG + PDG).

So, you can use the first project as a base for unified CFG.

**Comment by johndpope** [comment]
updates in this space https://github.com/src-d/code2vec - fyi @zurk

**Comment by zurk** [comment]
Hello, guys. Yes, we are implementing this article: https://arxiv.org/pdf/1803.09473.pdf using our own tooling like https://github.com/src-d/ml (for machine learning on the source code) and https://github.com/bblfsh/bblfshd/ to get Universal AST to be able to work with all languages in the same way.

What about google translator for the code it is a really cool idea and can be possible in some cases, but have a lot of underwater rocks. For example, we have `sort` function in many languages and in some of them you have `nan` value at the beginning of the list, in other cases in the end. Ok, have fun :) It can change your program behavior completely. 

P.S.: You guys have a really good tool!

**Comment by TensorTom** [comment]
I've also been investigating this possibility. Just came across this repo: https://github.com/pcyin/tranX

**Comment by johndpope** [comment]
related - 
natural language to executable code
https://github.com/pcyin/NL2code
https://arxiv.org/abs/1704.01696

fyi @pcyin

**Comment by inshua** [comment]
* backward access
* reference subtyped rule (\# marked rule)
* rule as set, support UNION, EXCLUDE and other operators

**Comment by KvanTTT** [comment]
Could you elaborate on these?

**Comment by inshua** [comment]
> @inshua Could you elaborate on these?

* backward access
* reference subtyped rule (# marked rule)
i.e.   VB support this syntax
```vb
For i = 1 To 10 
  For j = 1 To 10
    ...
Next j, i  ' close both j and i
```
I have solved it, but if ANTLR support backward reference it will be better
```antlr
nextStmt:
    NEXT    # OnlyNext
    | NEXT identifier   #NextId
    | nextStmt#NextId(-1) identifier  # NextIdMore   // backward1 is NextId
    | nextStmt#NextIdMore identifier
;
```
Here I show `reference subtyped rule` too, they are `nextStmt#NextId` and `nextStmt#NextIdMore`.

* rule as set, support UNION, EXCLUDE

we can write rule as 
```antlr
rule1 : (rule1 | rule2);
```
It's good, but if we treat rule as set, it just equivs `rule1 UNION rule2`, we should support `rule1 SUBTRACT rule2`, `rule1 AND rule2` too. i.e.  
```antlr
// wrong rules,  just for a presentation
multi: '*' | '/';
op : '+' | '-' | multi;
multiExpr: expr multi expr;
expr: multiExpr (op - multi) multiExpr;   // `op - multi` got '+' '-'
```
And `rule1 AND rule2` are very useful too.

**Comment by KvanTTT** [comment]
> rule as set, support UNION, EXCLUDE

Not sure such functionality should be integrated into ANTLR. It's out of EBNF and it's hard to imagine cases where it's required.

**Comment by inshua** [comment]
> > rule as set, support UNION, EXCLUDE
> 
> Not sure such functionality should be integrated into ANTLR. It's out of EBNF and it's hard to imagine cases where it's required.

yes, it's very useful, like subtyped rules. I'll post more cases when I met new.
